Source/JavaScriptCore/yarr/hasher.py

```diff
@@ -41,31 +74,189 @@ def stringHash(str):
+    def add64(a, b):
+        return (a + b) & mask64
+
+    def multi64(a, b):
+        return a * b & mask64
+
+    def wymum(A, B):
+        ha = A >> 32
+        hb = B >> 32
+        la = A & mask32
+        lb = B & mask32
+        rh = multi64(ha, hb)
+        rm0 = multi64(ha, lb)
+        rm1 = multi64(hb, la)
+        rl = multi64(la, lb)
+        t = add64(rl, (rm0 << 32))
+        c = int(t < rl)
+
+        lo = add64(t, (rm1 << 32))
+        c += int(lo < t)
+        hi = add64(rh, add64((rm0 >> 32), add64((rm1 >> 32), c)))
+        return lo, hi
+
+    def wymix(A, B):
+        A, B = wymum(A, B)
+        return A ^ B
+
+    def convert32BitTo64Bit(v):
+        v = (v | (v << 16)) & 0x0000_ffff_0000_ffff
+        return (v | (v << 8)) & 0x00ff_00ff_00ff_00ff
+
+    def convert16BitTo32Bit(v):
+        return (v | (v << 8)) & 0x00ff_00ff
+
+    charCount = len(string)
+    byteCount = charCount << 1
+    charIndex = 0
+    seed = 0
+    move1 = ((byteCount >> 3) << 2) >> 1
+
+    seed ^= wymix(seed ^ secret[0], secret[1])
+    a = 0
+    b = 0
+
+    def c2i(i):
+        return ord(string[i])
+
+    def wyr8(i):
+        v = c2i(i) | (c2i(i + 1) << 8) | (c2i(i + 2) << 16) | (c2i(i + 3) << 24)
+        return convert32BitTo64Bit(v)
+
+    def wyr4(i):
+        v = c2i(i) | (c2i(i + 1) << 8)
+        return convert16BitTo32Bit(v)
+
+    def wyr2(i):
+        return c2i(i) << 16
+
+    if byteCount <= 16:
+        if byteCount >= 4:
+            a = (wyr4(charIndex) << 32) | wyr4(charIndex + move1)
+            charIndex = charIndex + charCount - 2
+            b = (wyr4(charIndex) << 32) | wyr4(charIndex - move1)
+        elif byteCount > 0:
+            a = wyr2(charIndex)
+            b = 0
+        else:
+            a = b = 0
+    else:
+        i = byteCount
+        if i > 48:
+            see1 = seed
+            see2 = seed
+            while True:
+                seed = wymix(wyr8(charIndex) ^ secret[1], wyr8(charIndex + 4) ^ seed)
+                see1 = wymix(wyr8(charIndex + 8) ^ secret[2], wyr8(charIndex + 12) ^ see1)
+                see2 = wymix(wyr8(charIndex + 16) ^ secret[3], wyr8(charIndex + 20) ^ see2)
+                charIndex += 24
+                i -= 48
+                if i <= 48:
+                    break
+            seed ^= see1 ^ see2
+        while i > 16:
+            seed = wymix(wyr8(charIndex) ^ secret[1], wyr8(charIndex + 4) ^ seed)
+            i -= 16
+            charIndex += 8
+        move2 = i >> 1
+        a = wyr8(charIndex + move2 - 8)
+        b = wyr8(charIndex + move2 - 4)
+    a ^= secret[1]
+    b ^= seed
+
+    (a, b) = wymum(a, b)
+    hashValue = wymix(a ^ secret[0] ^ byteCount, b ^ secret[1]) & mask32
+
+    return finalizeAndMaskTop8Bits(hashValue)
+
+
+def ceilingToPowerOf2(v):
+    v -= 1
+    v |= v >> 1
+    v |= v >> 2
+    v |= v >> 4
+    v |= v >> 8
+    v |= v >> 16
+    v += 1
+    return v
+
+
+def createHashTable(keys, hashTableName, compactSize=0):
```

Comment:
This is used to create the `CompactHashIndex`s in `JSDollarVM.cpp`. Remove this or add a comment here?